### ISSUE TITLE
kt_android_library violates "strict deps"

### ISSUE THREAD

**Issue by donaldchai** [bazelbuild/rules_kotlin#331]
(Followup to https://github.com/bazelbuild/rules_kotlin/issues/152)

Due to this misleading error message:

https://github.com/bazelbuild/bazel/blob/f7d726f88f20c871d274b75a01f6a64d186403dc/src/main/java/com/google/devtools/build/lib/rules/android/AndroidLibrary.java#L58

many users of `android_library` rewrite their rules to use the `exports` attribute instead of `deps` to suppress the error.

Unfortunately, this is not always the correct resolution.  The fundamental problem is that `android_library` exports its `deps` when `srcs` is empty.  What the error message in AndroidLibrary should have really said is something like:

```
android_library without srcs exports its deps:
1) if you want to export the deps, move them to "exports".
2) if you don't want to export the deps, add a dummy source file to avoid that.
```

`kt_android_library` should use option <span>#</span>2.

**Comment by restingbull** [comment]
I think we'd be better off bringing it inline with java_library.exports.

**Comment by cgruber** [comment]
For java_library it barfs and tells you to turn them into runtime_deps.  I don't konw if these function like exports, but I don't think so.

**Comment by donaldchai** [comment]
> I think we'd be better off bringing it inline with java_library.exports.

I'm not sure what you mean.  exports are not deps, and deps are not exports.  They are completely unrelated.

https://github.com/bazelbuild/bazel/blob/df9825958128beaaa4c7ded4c433f276308dc5e0/src/main/java/com/google/devtools/build/lib/bazel/rules/java/BazelJavaLibraryRule.java#L112

**Comment by restingbull** [comment]
exports are most certainly a form of dep. They just happened to be a dep for the children not the library.

From the top, then:

There are main 3 attributes that contribute to the classpaths, as well as the runtime android resources and runfiles (and, please, do not get me started on the native/skylark runfiles integration)
1. deps
    * contributes to the library compile time classpath
    * contributes to the downstream compile classpath (strict deps is currently borked, as kotlin requires a different implementation of jdeps to be correct.)
    * contributes to the runtime classpath
    * contributes to the android resources, manifest, proguard, etc. etc. Non-strict by default
1. exports
    * contributes to the downstream compile classpath (strict deps is currently borked, as kotlin requires a different implementation of jdeps to be correct.)
    * contributes to the runtime classpath
    * contributes to the android resources, manifest, proguard, etc. etc. Non-strict by default
1. runtime_deps
    * contributes to the runtime classpath
    * contributes to the android resources, manifest, proguard, etc. etc. Non-strict by default
    * Do. Not. Ask. About. Neverlink. (the asymmetry of the neverlink and runtime deps allows the nonsensical runtime_deps = blah, neverlink = true, which has, depending on the rule, undefined behaviour)

The core issue is that the android_library used to, in Google3, provide exports if the srcs were omitted -- hence the warning which should have been copybara'd out, tbh. android_library didn't have exports till, idk, 3 years ago? And the resource handling is still pretty ambigious.

So, when a library with deps and no sources is defined we know that local compilation is not intended. There is no work to be down. That leads us to ask the awkward question about user intent:

Are they:
... trying to contribute compiletime and runtime classpaths?
... trying to contribute non-classpath artifacts?
... trying to contribute to the runtime classpath?
... did they make a mistake (probably through a macro)?

Suggesting that an empty source should be added is incredibly non-ideal as it:
* Creates new actions in the action graph. This is not free, especially when working in a resource constrained environment like a laptop.
* Fails to indicate rule intent
* Generates new edge cases for long term rule maintenance.

**Comment by donaldchai** [comment]
100% agree that usage should reflect the developer's intent.  The intent of this:

https://github.com/bazelbuild/rules_kotlin/blob/d83e2689b04941d097be7d3b85fd227434e036ea/kotlin/internal/jvm/android.bzl#L28

is to perform resource processing, including generating R classes.  The intent is *not* to export dependencies, but this intent is now lost:

https://github.com/bazelbuild/rules_kotlin/commit/c5c3503c8ca447a285c4ec38e3483e9cadf64058#diff-8e550ff58ac97ec304a6b5e265322855

It's unfortunate that android_library has the set of semantics that it does (deps-versus-exports is one of many problems).  Fixing those would likely entail a painful incompatible-change process.  I would recommend not exacerbating these same bugs in kt_android_library's semantics to reduce the migration pain.
